sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/properties/AadAuthenticationProperties.java

```diff
@@ -552,100 +568,115 @@ private void validateApplicationType() {
-                            "resource_server", registrationId, "authorization_code"));
-                    }
-                    if (ON_BEHALF_OF.getValue().equals(grantType)) {
-                        throw new IllegalStateException(String.format(UNMATCHING_OAUTH_GRANT_TYPE_FROMAT,
-                            "resource_server", registrationId, "on_behalf_of"));
-                    }
-                    break;
-                case RESOURCE_SERVER_WITH_OBO:
-                    if (AUTHORIZATION_CODE.getValue().equals(grantType)) {
-                        throw new IllegalStateException(String.format(UNMATCHING_OAUTH_GRANT_TYPE_FROMAT,
-                            "resource_server_with_obo", registrationId, "authorization_code"));
-                    }
-                    break;
-                case WEB_APPLICATION_AND_RESOURCE_SERVER:
-                default:
-                    LOGGER.debug("'spring.cloud.azure.active-directory.authorization-clients.{}.authorization-grant-type' is valid.", registrationId);
-            }
+        if (grantType == null) {
+            grantType = decideDefaultGrantTypeFromApplicationType(registrationId, applicationType);
+            properties.setAuthorizationGrantType(grantType);
+        }
+
+        // Validate authorization grant grantType
+        validateAuthorizationGrantType(registrationId, grantType, properties);
+
+        // Extract validated scopes from properties
+        List<String> scopes = extractValidatedScopes(registrationId, properties);
+        addNecessaryScopesForAuhtorizationCodeClients(properties, scopes);
+    }
 
-            if (AZURE_CLIENT_REGISTRATION_ID.equals(registrationId)
-                && AUTHORIZATION_CODE != properties.getAuthorizationGrantType()) {
-                throw new IllegalStateException("spring.cloud.azure.active-directory.authorization-clients."
-                    + AZURE_CLIENT_REGISTRATION_ID
-                    + ".authorization-grant-type must be configured to 'authorization_code'.");
+    /**
+     * Add necessary scopes for authorization_code clients.
+     *
+     * https://docs.microsoft.com/graph/permissions-reference#remarks-17
+     * https://docs.microsoft.com/azure/active-directory/develop/v2-permissions-and-consent#openid-connect-scopes
+     *
+     * "openid" : allows to request an ID token.
+     * "profile" : allows returning additional claims in the ID token.
+     * "offline_access" : allows to request a refresh token.
+     * @param properties AuthorizationClientProperties
+     * @param scopes scopes for authorization_code clients.
+     */
+    private void addNecessaryScopesForAuhtorizationCodeClients(AuthorizationClientProperties properties,
+                                                               List<String> scopes) {
+        if (properties.getAuthorizationGrantType().equals(AUTHORIZATION_CODE)) {
+            String[] scopesNeeded = new String[] { "openid", "profile", "offline_access" };
```

Comment:
put this as a `static` field

Reply:
Let's keep it in local scope, static does not make sense.